Add resources for creating ML managed alerts

csmarchbanks opened the pull request.

csmarchbanks committed:
Add resources for creating ML managed alerts

julienduchesne commented:
Same comment as https://github.com/grafana/terraform-provider-grafana/pull/1653#issuecomment-2179294807 😓

csmarchbanks committed:
Refactor resource_alert.go to use framework

csmarchbanks committed:
Specifically handle nil map case

csmarchbanks committed:
Add default values for string fields

csmarchbanks commented:
Ok, updated to use the new framework and got tests passing again.

julienduchesne commented on examples/resources/grafana_machine_learning_alert/job_alert.tf:
```diff
@@ -0,0 +1,17 @@
+resource "grafana_machine_learning_job" "test_alert_job" {
```
Can you rename one of the two examples to `resource.tf` so that it shows up in the docs?

csmarchbanks replied:
Done!

julienduchesne commented on internal/resources/machinelearning/resource_alert_test.go:
```diff
@@ -0,0 +1,156 @@
+					testAccMLOutlierCheckExists("grafana_machine_learning_outlier_detector.test_alert_outlier_detector", &outlier),
+					testAccMLOutlierAlertCheckExists("grafana_machine_learning_alert.test_outlier_alert", &outlier, &alert),
+				),
+			},
```
I would add an import step to both of these tests to make sure that reads work 100% correctly

julienduchesne commented on internal/resources/machinelearning/resource_alert.go:
```diff
@@ -0,0 +1,392 @@
+				Description: "How long values must be anomalous before firing an alert.",
+				Optional:    true,
+				Computed:    true,
+				Default:     stringdefault.StaticString("0s"),
```
I would add a validator for those duration strings, so that they're caught in a plan and not in the apply stage

csmarchbanks replied:
Done as well.

julienduchesne reviewed (commented):
Nice 🚀  A few small things to fix

csmarchbanks committed:
Add duration validation and test

csmarchbanks committed:
Rename job_alert.tf to resource.tf

This generates an example in the documentation for how to create an
alert resource.

csmarchbanks committed:
Implement importing of ML alerts

csmarchbanks commented:
Addressed all the comments but I am running into an OnCall test that seems to be consistently failing:
```
=== NAME  TestAccDataSourceTeam_Basic
    data_source_team_test.go:17: Step 1/1 error: Error running apply: exit status 1
        
        Error: couldn't find a team matching: test-acc-pgbfl46w
        
          with data.grafana_oncall_team.test-acc-team,
          on terraform_plugin_test.tf line 6, in data "grafana_oncall_team" "test-acc-team":
           6: data "grafana_oncall_team" "test-acc-team" {
        
--- FAIL: TestAccDataSourceTeam_Basic (62.65s)
```

edit: enough retries and everything passed!

ioanarm commented on internal/resources/machinelearning/resource_alert.go:
```diff
@@ -0,0 +1,478 @@
+		return
+	}
+	if data == nil {
+		resp.Diagnostics.AddError("Resource not found", "Resource not found")
```
maybe `resource` instead of the capital for consistency

csmarchbanks replied:
I changed most of the other ones to consistently use sentence case since this is a user facing error.

ioanarm commented on internal/resources/machinelearning/resource_alert.go:
```diff
@@ -0,0 +1,478 @@
+		alert, err = r.mlapi.NewOutlierAlert(ctx, data.OutlierID.ValueString(), alert)
+	}
+	if err != nil {
+		resp.Diagnostics.AddError("Unable to Create Resource", err.Error())
```
same

ioanarm commented on internal/resources/machinelearning/resource_alert.go:
```diff
@@ -0,0 +1,478 @@
+		return
+	}
+	if readData == nil {
+		resp.Diagnostics.AddError("Unable to read created resource", "Resource not found")
```
same as above

ioanarm reviewed (approved):
Looks good to me Chris, the examples as well. 
Had a look at the new resources to try and make sense of the changes. 

Let's wait for @julienduchesne to give it another go though in case there is something else though.

csmarchbanks committed:
Use sentence case for all error diagnostics
